slack-conversation.c

```diff
@@ -35,10 +36,111 @@ static void conversations_list_cb(SlackAccount *sa, gpointer data, json_value *j
 		slack_login_step(sa);
 }
 
+static void get_latest_history_cb(SlackAccount *sa, gpointer data, json_value *json, const char *error) {
+	json_value *chan_json = json_get_prop(json, "channel");
+	if (!chan_json)
+		return;
+
+	SlackObject *conv;
+	slack_object_id id;
+	const char *sid;
+
+	sid = json_get_prop_strptr(chan_json, "user");
+	if (sid) {
+		// Direct IM, use user ID.
+		slack_object_id_set(id, sid);
+		conv = g_hash_table_lookup(sa->users, id);
+	} else {
+		// MPIM, use conversation ID.
+		sid = json_get_prop_strptr(chan_json, "id");
+		if (!sid) {
+			purple_debug_warning("slack", "slack response did not include 'user' or 'id' while fetching history\n");
+			return;
+		}
+		slack_object_id_set(id, sid);
+		conv = g_hash_table_lookup(sa->channels, id);
+	}
+
+	if (!conv) {
+		purple_debug_warning("slack", "unable to locate user/channel '%s' while fetching history\n", sid);
+		return;
+	}
+
+	purple_debug_misc("slack", "fetching history for conversation '%s'\n", sid);
+	slack_get_history(sa, conv,
+			  json_get_prop_strptr(chan_json, "last_read"),
+			  json_get_prop_val(chan_json, "unread_count", integer, 0));
+}
+
+static gboolean get_queued_unread_messages_cb(gpointer data) {
+	SlackAccount *sa = data;
+
+	if (!g_queue_is_empty(sa->fetch_unread_queue)) {
+		char *chan_id = g_queue_pop_head(sa->fetch_unread_queue);
+		slack_api_call(sa, get_latest_history_cb, NULL, "conversations.info", "channel", chan_id, NULL);
+		g_free(chan_id);
+	}
+
+	if (!g_queue_is_empty(sa->fetch_unread_queue)) {
+		/* Wait and then fetch next. */
+		return TRUE;
+	} else {
+		/* We won't need the queue anymore. */
+		g_queue_free(sa->fetch_unread_queue);
+		sa->fetch_unread_queue = NULL;
+		sa->fetch_unread_timer = 0;
+		return FALSE;
+	}
+}
+
+static void get_unread_messages_cb(SlackAccount *sa, gpointer data, json_value *json, const char *error) {
+	json_value *all_ims[2] = { json_get_prop(json, "ims"), json_get_prop(json, "mpims") };
+
+	if (!sa->fetch_unread_queue) {
+		sa->fetch_unread_queue = g_queue_new();
+	}
+
+	for (unsigned im_type = 0; im_type < 2; im_type++) {
+		json_value *chans = all_ims[im_type];
+		for (unsigned i = 0; chans && i < chans->u.array.length; i++) {
+			json_value *chan_json = chans->u.array.values[i];
+			const char *latest = json_get_prop_strptr(chan_json, "latest");
+			const char *last_read = json_get_prop_strptr(chan_json, "last_read");
+			if (latest && last_read && strcmp(latest, last_read) == 0) {
+				/* Skip API call if we have read everything already. */
+				continue;
+			}
+			const char *chan_id = json_get_prop_strptr(chan_json, "id");
+			if (!chan_id) {
+				purple_debug_warning("slack", "slack response did not include 'id' while fetching unread messages\n");
+				continue;
+			}
+
+			g_queue_push_tail(sa->fetch_unread_queue, g_strdup(chan_id));
+		}
+	}
+
+	if (!g_queue_is_empty(sa->fetch_unread_queue)) {
+		sa->fetch_unread_timer = purple_timeout_add_seconds(5, get_queued_unread_messages_cb, sa);
```

Comment:
These called should be serialized, not just staggered. 5 seconds can be a very long time for large slacks, or not long enough if previous calls haven't finished or are throttled.